doc/ax/ax.txt

```diff
@@ -0,0 +1,2095 @@
+(what it's processing) may change what functionality is available and, more
+importantly, the behaviour of reading/writing from attributes. See the @ref axexecutionxcontext
+for details.
+@par
+For details of extending AX for custom geometry, see the @subpage openvdbaxapi
+developer documentation. Note that custom geometry support requires C++ extension.
+
+
+@subsection axprogramexample A Program Example
+@par
+Before getting into the technical depths of the language, it's a good idea to
+observe a small but complete AX program which could be compiled and executed.
+To begin with, here is a very simple example of a program that reads and writes
+to a particular attribute:
+@par
+@code{.c}
+// read a floating point attribute from the value of "myattribute"
+float temp = float@myattribute;
+// if the value is less than 0, clamp it to 0
+if (temp < 0.0f) float@myattribute = 0.0f;
+@endcode
+Note that there is no other required logic here to make this a compatible
+program i.e. feeding this code to the @ref vdbaxbinary "vdb_ax command line binary"
+will compile and execute over provided OpenVDB files.
+@par
+The following explains the above example in relation to OpenVDB data; OpenVDB
+points and OpenVDB volumes. The example demonstrates reading from a value on
+some input, either a point attribute or a voxel value, and storing the
+result in a @ref axdecls "local variable". Importantly, the attribute being
+accessed has the name @b `myattribute` and the type @b `float`. The first statement:
+@code{.c}
+float temp = float@myattribute;
+@endcode
+Invokes a @b read from this attribute (see @ref axattribaccess), retrieving the
+value from the geometry and storing it in the variable @b `temp` to be used in
+the AX program. The second statement:
+@code{.c}
+if (temp > 0.0f) float@myattribute = 0.0f;
+@endcode
+Performs a comparison of this value with @b `0.0f` and, if the value is greater
+than @b `0.0f`, performs a @b write to the geometry in the form of
+<b>`float@myattribute = 0.0f;`</b> which sets the value of this attribute on the
+geometry to @b `0.0f`.
+@par
+For OpenVDB Points, this kernel is run over every point in an OpenVDB Points
+Grid. In simple pseudo code:
+@code{.unparsed}
+foreach(point in pointgrid)
+    run_ax_kernel(point)
+done()
+@endcode
+Where @b `pointgrid` is a single Point Data Grid. For OpenVDB volumes the kernel
+is run over each voxel in a provided grid:
+@code{.unparsed}
+foreach(grid in grids)
+    foreach(voxel IN grid)
+        run_ax_kernel(voxel)
+    done()
+done()
+@endcode
+@par
+Where @b `grids` can comprise of any number of OpenVDB volumes. In this program,
+only a floating point grid with the name @b `myattribute` will be updated. Each
+voxel in this grid will have it's value compared in the same way as the points
+example given above.
+
+
+@subsection axexecutionxcontext OpenVDB Execution Context
+@par
+AX programs are designed to run over different types of geometry, specifically
```

Comment:
"types of geometry" -> "types of grids". While points could be referred to as geometry I don't think it's a good description for voxels

Reply:
Done